mass2/core/channels.py

```diff
@@ -653,3 +656,81 @@ def from_df(
                 npulses=len(df),
             )
         return Channels(channels, description)
+
+    def save_analysis(self, path: Path | str, overwrite: bool = False) -> None:
+        """Save an analysis-in-progress completely to a directory
+
+        Parameters
+        ----------
+        path : Path | str
+            Directory to save work in. If it doesn't exist, its parent should.
+        overwrite : bool, optional
+            If `path` exists, whether to overwrite it, by default False
+        """
+        path = pathlib.Path(path)
+        if path.suffix != ".zip":
+            path = path.with_suffix(".zip")
+
+        if os.path.exists(path) and not overwrite:
+            raise ValueError(f"Set save_analysis(...overwrite=True) to overwrite the existing {path=}")
+
+        def drop_columns(df: pl.DataFrame, *columns: str) -> pl.DataFrame:
+            columns_to_drop = set(columns)
+            columns_to_drop = columns_to_drop.intersection(df.columns)
+            for c in columns_to_drop:
+                df = df.drop(c)
+            return df
+
+        with ZipFile(str(path), "w") as zf:
+            channels = {}
+            bad_channels = {}
+            for ch_num, ch in self.channels.items():
+                df = drop_columns(ch.df, "pulse", "timestamp", "subframecount")
+                buffer = io.BytesIO()
+                df.write_parquet(buffer)
+                fname = f"data_chan{ch_num:04d}.parquet"
+                zf.writestr(fname, buffer.getvalue())
+                steps = ch.steps.trim_dead_ends(None, drop_debug=True)
+                channels[ch_num] = dataclasses.replace(ch, df=pl.DataFrame(), df_history=[], noise=None, steps=steps)
+            for ch_num, badch in self.bad_channels.items():
+                df = drop_columns(badch.ch.df, "pulse", "timestamp", "subframecount")
+                buffer = io.BytesIO()
+                df.write_parquet(buffer)
+                fname = f"data_bad_chan{ch_num:04d}.parquet"
+                zf.writestr(fname, buffer.getvalue())
+                steps = badch.ch.steps.trim_dead_ends(None, drop_debug=True)
```

Comment:
Every time I read this I worry you're dropping dead ends, but you aren't. I suggest making a simple function so you can call `ch.steps.drop_debug()` so it's more clear what the code is doing.

Reply:
Okay. Done.